### PR TITLE
fix(crypto): crc32 algorithm fix

### PR DESCRIPTION
**Pull request by dariaterekhova-actionengine**
During zip functionality development I found out that our crc32 algorithm returns wrong values. So I made a fix and checked on a several files that hashes now are exactly the same as in bash zip utility.

### CODE REVIEW

**Comment by ibgreen** [comment]
Any references that you can link to in code? More test cases?

File: modules/crypto/test/crypto.spec.ts [hunk 1]
```diff
@@ -34,13 +41,17 @@ const TEST_CASES = [
     title: 'binary data (repeated)',
     data: repeatedData,
     digests: {
-      sha256: 'SnGMX2AgkPh21d2sxow8phQa8lh8rjf2Vc7GFCIwj2g='
-      // 'bSCTuOJei5XsmAnqtmm2Aw/2EvUHldNdAxYb3mjSK9s=',
+      sha256: 'bSCTuOJei5XsmAnqtmm2Aw/2EvUHldNdAxYb3mjSK9s='
```
[the remark this answers is not in the document]
Comment (dariaterekhova-actionengine) [reply modules/crypto/test/crypto.spec.ts:44]: I checked it with `shasum` bash utility and the one that had been commented is the correct one for this data. I found out that some of our crypto classes gives the correct result only once after the initialisation. But this test had been used the same class multiple times, so it has given the wrong result on second use and after. So probably somebody wanted just to make this test work...

File: modules/crypto/src/lib/algorithms/crc32.ts [hunk 2]
```diff
@@ -24,6 +25,7 @@ export default class CRC32 {
   }
 
   finalize() {
+    // performing final division and making crc unsigned.
     this.crc = (this.crc ^ -1) >>> 0;
```
[the remark this answers is not in the document]
Comment (dariaterekhova-actionengine) [reply modules/crypto/src/lib/algorithms/crc32.ts:29]: The difference there that 1 in bite representation is `00000000000000000000000000000001` and -1 is `11111111111111111111111111111111`. So `Math.abs()` completely changes the bits there, but what we need is just make JS count first `1` bit as a part of a number and not the sign marker. We don't need to change anything else in this bit representation. And that's exactly what `>>>` does.

We can also implement it like this:
```
const toUnsignedInt32 = (n: number): number => {
  if (n >= 0) {
    return n;
  }
  return 0xFFFFFFFF - (n * -1) + 1;
}
```
We perform this operation once per hash generation, so it shouldn't affect performance

File: modules/crypto/test/crypto.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/crypto/test/crypto.spec.ts:44]: That is a huge bug that we must fix first. We shouldn't work around it! 

https://github.com/visgl/loaders.gl/pull/2824

File: modules/crypto/test/crypto.spec.ts [hunk 3]
```diff
@@ -34,13 +41,17 @@ const TEST_CASES = [
...
     }
   }
 ];
 
-const HASHES = [new CRC32Hash(), new CRC32CHash(), new MD5Hash(), new SHA256Hash({modules})];
+const HASHES = {
+  crc32: () => new CRC32Hash(),
```
[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/crypto/test/crypto.spec.ts:50]: Nope, fixing in https://github.com/visgl/loaders.gl/pull/2824

**Review by ibgreen — COMMENTED** [review]
@dariaterekhova-actionengine I landed the hash reuse fixes. If you rebase on master and remove any code that was working around that issue, this should be good to land.

**Review by ibgreen — APPROVED** [review]
Looks great, thanks for good teamwork!

**Pull request merged**